docs/reference/advanced/baggage.md

```diff
@@ -0,0 +1,90 @@
+# Baggage
+
+In OpenTelemetry, [Baggage](https://opentelemetry.io/docs/concepts/signals/baggage/) is a key-value store that can be
+used to propagate any data you like alongside context during (distributed) tracing.
+Baggage is included in the context of a trace, and is propagated across service boundaries, allowing you to
+attach metadata to a trace that can be accessed by any service that participates in that trace.
+
+Technically, Baggage is a separate key-value store and is formally unassociated with attributes on spans, metrics, or
+logs without explicitly creating those attributes from the Baggage.
+
+However, a common use cases for Baggage is to add data to span attributes across a whole trace. Because of this,
+**Logfire** provides convenience APIs for setting Baggage and automatically adding Baggage to descendant span
+attributes.
+
+## Why use Baggage to set attributes?
+
+Though tracing makes it a lot easier to understand and visualize the relationship between parent and child spans,
+it's still usually easiest to express filtering criteria on an _individual_ span when doing custom searches,
+aggregations, etc.
+
+As a concrete example, you might have received an issue report from a user with user_id `scolvin`, and want to
+search for all spans corresponding to slow database queries that happened in requests made by that user. You might find
+all slow database requests with the query `duration > 1 AND attributes ? 'db.statement'`, but trying to filter
+down to only those spans coming from requests by a specific user is harder.
+
+## Basic usage
+
+Here's how to solve the above example using Baggage:
+
+```python
+import logfire
+
+logfire.configure()
+
+with logfire.set_baggage(user_id='scolvin'):
+    print(logfire.get_baggage())  # (just for demonstration, not usually needed)
+    #> {'user_id': 'scolvin'}
+
+    # All spans opened here (and their descendants)
+    # will have the attribute `user_id` set to 'scolvin'
+    # (or more attributes if you set other baggage values)
+    with logfire.span('inside-baggage-span'):
+        ...
+```
+
+Then you can update your query to
+`duration > 1 AND attributes ? 'db.statement' AND attributes->>'user_id' = 'scolvin'`.
+
+## Disabling
+
+If you don't want to add Baggage to span attributes, pass `add_baggage_to_attributes=False` to the
+[`logfire.configure()`][logfire.configure] function. This may slightly improve performance.
+The [`set_baggage`][logfire.set_baggage] contextmanager will still update the OpenTelemetry Baggage and propagate it to other services.
+
+## Baggage values requirements
+
+While baggage can technically contain any data type, only strings will be set as attributes on spans.
```

Comment:
I don't feel like this is the right direction to give users. I think we should be explicit that baggage can only contain strings because it is serialized across the wire and we should enforce that in type hints and runtime assertions / errors / warnings (I'm okay with doing best effort to dump non-strings to strings but I think that should basically be undocumented and emit a warning).